attestation/secretscan/secretscan.go

```diff
@@ -0,0 +1,449 @@
+	subjects        map[string]cryptoutil.DigestSet
+}
+
+func New(opts ...Option) *Attestor {
+	a := &Attestor{
+		scanBinaries:    defaultScanBinaries,
+		failOnDetection: defaultFailOnDetection,
+		subjects:        make(map[string]cryptoutil.DigestSet),
+	}
+
+	for _, opt := range opts {
+		opt(a)
+	}
+
+	return a
+}
+
+func (a *Attestor) Name() string {
+	return Name
+}
+
+func (a *Attestor) Type() string {
+	return Type
+}
+
+func (a *Attestor) RunType() attestation.RunType {
+	return RunType
+}
+
+func (a *Attestor) Schema() *jsonschema.Schema {
+	return jsonschema.Reflect(a)
+}
+
+func (a *Attestor) Attest(ctx *attestation.AttestationContext) error {
+	// Create a temporary directory for scanning
+	tempDir, err := os.MkdirTemp("", "secretscan")
+	if err != nil {
+		return fmt.Errorf("error creating temp dir: %w", err)
+	}
+	defer os.RemoveAll(tempDir)
+
+	// Initialize Gitleaks detector
+	detector, err := a.initGitleaksDetector()
+	if err != nil {
+		return fmt.Errorf("error initializing gitleaks detector: %w", err)
+	}
+
+	// Scan attestations
+	if err := a.scanAttestations(ctx, tempDir, detector); err != nil {
+		log.Debugf("(attestation/secretscan) error scanning attestations: %s", err)
+	}
+
+	// Scan products if enabled
+	if a.scanBinaries {
+		if err := a.scanProducts(ctx, tempDir, detector); err != nil {
+			log.Debugf("(attestation/secretscan) error scanning products: %s", err)
+		}
+	}
+
+	// If configured to fail on detection and we found secrets, return an error
+	if a.failOnDetection && len(a.Findings) > 0 {
+		return fmt.Errorf("secret scanning failed: found %d secrets", len(a.Findings))
+	}
+
+	return nil
+}
+
+func (a *Attestor) initGitleaksDetector() (*detect.Detector, error) {
+	// Create a new detector with default config
+	detector, err := detect.NewDetectorDefaultConfig()
+	if err != nil {
+		return nil, fmt.Errorf("error creating gitleaks detector: %w", err)
+	}
+	return detector, nil
+}
+
+func (a *Attestor) scanAttestations(ctx *attestation.AttestationContext, tempDir string, detector *detect.Detector) error {
+	// Get all completed attestors
+	completedAttestors := ctx.CompletedAttestors()
+
+	// Process each attestor
+	for _, completed := range completedAttestors {
+		// Skip scanning ourselves
+		if completed.Attestor.Name() == Name {
+			continue
+		}
+
+		// Convert attestor to JSON for scanning
+		attestorJSON, err := json.MarshalIndent(completed.Attestor, "", "  ")
+		if err != nil {
+			log.Debugf("(attestation/secretscan) error marshaling attestor %s: %s", completed.Attestor.Name(), err)
+			continue
+		}
+
+		// Create a file for the attestor content
+		filename := filepath.Join(tempDir, fmt.Sprintf("attestor_%s.json", completed.Attestor.Name()))
+		if err := os.WriteFile(filename, attestorJSON, 0644); err != nil {
+			log.Debugf("(attestation/secretscan) error writing temp file: %s", err)
+			continue
+		}
+
+		// Scan the file with Gitleaks
+		findings, err := a.ScanFile(filename, detector)
```

Comment:
I'm worried calling `ScanFile` for every file in the product list may be incredibly slow. Was using `[DetectDirectory](https://github.com/gitleaks/gitleaks/blob/af7d5bcb83255283767c45e03c904af9a77e6e75/detect/directory.go#L20)` not an option here? It looks like it uses some concurrency, and while it may still be slow it may not be as slow as doing all section in serial.

Reply:
Me too.  I'll remove it for now